pkg/idp/service.go

```diff
@@ -0,0 +1,327 @@
+package idp
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/canonical/identity-platform-admin-ui/internal/logging"
+	"github.com/canonical/identity-platform-admin-ui/internal/monitoring"
+	"github.com/google/uuid"
+	"go.opentelemetry.io/otel/trace"
+	"gopkg.in/yaml.v3"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	coreV1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+type Config struct {
+	Name      string
+	Namespace string
+	KeyName   string
+	K8s       coreV1.CoreV1Interface
+}
+
+type Service struct {
+	cmName      string
+	cmNamespace string
+	keyName     string
+
+	k8s coreV1.CoreV1Interface
+
+	tracer  trace.Tracer
+	monitor monitoring.MonitorInterface
+	logger  logging.LoggerInterface
+}
+
+func (s *Service) ListResources(ctx context.Context) ([]*Configuration, error) {
+	ctx, span := s.tracer.Start(ctx, "idp.Service.ListResources")
+	defer span.End()
+
+	cm, err := s.k8s.ConfigMaps(s.cmNamespace).Get(ctx, s.cmName, metaV1.GetOptions{})
+
+	if err != nil {
+		s.logger.Error(err.Error())
+		return nil, err
+	}
+
+	return s.idpConfiguration(cm.Data), nil
+
+}
+
+func (s *Service) GetResource(ctx context.Context, providerID string) ([]*Configuration, error) {
+	ctx, span := s.tracer.Start(ctx, "idp.Service.GetResource")
+	defer span.End()
+
+	cm, err := s.k8s.ConfigMaps(s.cmNamespace).Get(ctx, s.cmName, metaV1.GetOptions{})
+
+	if err != nil {
+		s.logger.Error(err.Error())
+		return nil, err
+	}
+
+	idps := s.idpConfiguration(cm.Data)
+
+	if idps == nil {
+		return nil, nil
+	}
+
+	// TODO @shipperizer find a better way to index the idps
+	for _, idp := range idps {
+		if idp.ID == providerID {
+			return []*Configuration{idp}, nil
+		}
+	}
+	return []*Configuration{}, nil
+}
+
+func (s *Service) EditResource(ctx context.Context, providerID string, data *Configuration) ([]*Configuration, error) {
+	ctx, span := s.tracer.Start(ctx, "idp.Service.EditResource")
+	defer span.End()
+
+	cm, err := s.k8s.ConfigMaps(s.cmNamespace).Get(ctx, s.cmName, metaV1.GetOptions{})
+
+	if err != nil {
+		s.logger.Error(err.Error())
+		return nil, err
+	}
+
+	idps := s.idpConfiguration(cm.Data)
+
+	if idps == nil {
+		return nil, nil
+	}
+
+	var idp *Configuration
+	// TODO @shipperizer find a better way to index the idps
+	for _, i := range idps {
+		if i.ID == providerID {
+			i = s.mergeConfiguration(i, data)
+			idp = i
+		}
+	}
```

Comment:
You could use the provider ID as the key in the configMap, you can switch from:
```yaml
idps.yaml:
  - provider_id: provider_1
    ...
  - provider_id: provider_2
    ...
  - provider_id: provider_3
    ...
```
To:
```yaml
provider_1:
  provider_id: provider_1
  ...
provider_2:
  provider_id: provider_2
  ...
provider_3:
  provider_id: provider_3
  ...
```
This way you should be able to fetch the provider simply by indexing the map. Is there a problem with this?